// @zola/app/styles/chat.message.styles.ts
import { StyleSheet } from 'react-native';

export const chatMessageStyles = StyleSheet.create({
	bubbleContainer: {
		position: 'relative',
	},
	messagePressable: {
		borderRadius: 16,
	},
	hoverActionsContainer: {
		position: 'absolute',
		top: '50%',
		flexDirection: 'row',
		gap: 6,
		zIndex: 200,
		backgroundColor: 'transparent',
		transform: [{ translateY: -18 }],
	},
	hoverActionsLeft: {
		left: -66,
	},
	hoverActionsRight: {
		right: -66,
	},
	hoverActionButton: {
		width: 36,
		height: 36,
		borderRadius: 18,
		backgroundColor: '#2C2C2E',
		justifyContent: 'center',
		alignItems: 'center',
		shadowColor: '#000',
		shadowOffset: { width: 0, height: 2 },
		shadowOpacity: 0.25,
		shadowRadius: 3,
		elevation: 5,
	},
	hoverActionText: {
		fontSize: 18,
		color: '#FFFFFF',
	},
	hoverActionEmoji: {
		fontSize: 18,
	},
	reactionsContainer: {
		flexDirection: 'row',
		flexWrap: 'wrap',
		marginTop: 4,
	},
	reactionsContainerLeft: {
		marginLeft: 50,
	},
	reactionsContainerRight: {
		marginLeft: 0,
	},
	reactionChip: {
		flexDirection: 'row',
		alignItems: 'center',
		backgroundColor: '#FFFFFF',
		borderRadius: 12,
		paddingHorizontal: 8,
		paddingVertical: 4,
		marginRight: 4,
		marginBottom: 4,
		borderWidth: 1,
		borderColor: '#E0E0E0',
	},
	reactionChipEmoji: {
		fontSize: 14,
		marginRight: 4,
	},
	reactionChipCount: {
		fontSize: 12,
		color: '#666666',
	},
	editedLabel: {
		fontSize: 11,
		marginTop: 4,
		fontStyle: 'italic',
	},
	editedLabelLeft: {
		color: '#666666',
		marginLeft: 50,
	},
	editedLabelRight: {
		color: 'rgba(255,255,255,0.7)',
		marginLeft: 0,
	},
reactionOverlay: {
		position: 'absolute',
		top: 0,
		left: 0,
		right: 0,
		bottom: 0,
		zIndex: 400,
		backgroundColor: 'transparent',
	},
	reactionBarWrapper: {
		position: 'absolute',
		zIndex: 401,
		transform: [{ translateX: -90 }],
	},
	reactionBarContainer: {
		flexDirection: 'row',
		backgroundColor: '#1C1C1E',
		borderRadius: 24,
		paddingHorizontal: 10,
		paddingVertical: 6,
		alignItems: 'center',
		shadowColor: '#000',
		shadowOffset: { width: 0, height: 4 },
		shadowOpacity: 0.3,
		shadowRadius: 6,
		elevation: 8,
	},
	reactionBarContainerIncoming: {},
	reactionBarContainerOwn: {},
	reactionBarTail: {
		width: 12,
		height: 12,
		backgroundColor: '#1C1C1E',
		position: 'absolute',
		bottom: -4,
		transform: [{ rotate: '45deg' }],
		borderRadius: 2,
	},
	reactionEmojiButton: {
		width: 36,
		height: 36,
		borderRadius: 18,
		backgroundColor: '#2C2C2E',
		justifyContent: 'center',
		alignItems: 'center',
		marginHorizontal: 4,
	},
	reactionEmojiText: {
		fontSize: 20,
	},
	reactionEmojiAddButton: {
		width: 36,
		height: 36,
		borderRadius: 18,
		backgroundColor: '#3A3A3C',
		justifyContent: 'center',
		alignItems: 'center',
		marginLeft: 4,
	},
	reactionEmojiAddText: {
		fontSize: 20,
		color: '#FFFFFF',
		fontWeight: '600',
	},
	// Inline bar (above bubble)
	reactionBarInline: {
		position: 'absolute',
		top: -56,
		zIndex: 300,
		flexDirection: 'row',
		backgroundColor: '#1C1C1E',
		borderRadius: 24,
		paddingHorizontal: 10,
		paddingVertical: 6,
		alignItems: 'center',
		shadowColor: '#000',
		shadowOffset: { width: 0, height: 4 },
		shadowOpacity: 0.3,
		shadowRadius: 6,
		elevation: 8,
	},
	reactionBarInlineLeft: { left: 0 },
	reactionBarInlineRight: { right: 0 },
});
